c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -2749,7 +2749,8 @@ class DeviceCachingAllocator {
         ? CUDAAllocatorConfig::max_split_size()
         : key.size;
     auto it = pool.blocks.lower_bound(&key);
-    if (it == pool.blocks.end() || (*it)->stream != p.stream()) {
+    if (it == pool.blocks.end() || (*it)->stream != p.stream() ||
+        (*it)->expandable_segment_) {
```

Comment:
Technically, couldn't there be a block that is big enough, but the biggest block is expandable segment and so we picked out the wrong one in the lower_bound test?

Reply:
Yeah, but in that case, it will just go to the else statement and free that single block anyway.